To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News


Unsloth's [Docker image](https://hub.docker.com/r/unsloth/unsloth) is here! Start training with no setup & environment issues. [Read our Guide](https://docs.unsloth.ai/new/how-to-train-llms-with-unsloth-and-docker).

[gpt-oss RL](https://docs.unsloth.ai/new/gpt-oss-reinforcement-learning) is now supported with the fastest inference & lowest VRAM. Try our [new notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/gpt-oss-(20B)-GRPO.ipynb) which creates kernels!

Introducing [Vision](https://docs.unsloth.ai/new/vision-reinforcement-learning-vlm-rl) and [Standby](https://docs.unsloth.ai/basics/memory-efficient-rl) for RL! Train Qwen, Gemma etc. VLMs with GSPO - even faster with less VRAM.

Unsloth now supports Text-to-Speech (TTS) models. Read our [guide here](https://docs.unsloth.ai/basics/text-to-speech-tts-fine-tuning).

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [6]:
%%capture
import os, re
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth
else:
    # Do this only in Colab notebooks! Otherwise use pip install unsloth
    import torch; v = re.match(r"[0-9\.]{3,}", str(torch.__version__)).group(0)
    xformers = "xformers==" + ("0.0.32.post2" if v == "2.8.0" else "0.0.29.post3")
    !pip install --no-deps bitsandbytes accelerate {xformers} peft trl triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf "datasets>=3.4.1,<4.0.0" "huggingface_hub>=0.34.0" hf_transfer
    !pip install --no-deps unsloth
!pip install transformers==4.55.4
!pip install --no-deps trl==0.22.2

In [7]:
%%capture
!pip install --no-deps git+https://github.com/huggingface/transformers.git unsloth_zoo# Need main branch for Liquid LFM2 models
!pip install --no-deps causal-conv1d==1.5.0.post8 # Install Mamba kernels

### Unsloth

In [8]:
from unsloth import FastModel
import torch

fourbit_models = [
    # 4bit dynamic quants for superior accuracy and low memory use
    "unsloth/LFM2-1.2B-unsloth-bnb-4bit",
    "unsloth/LFM2-700M-unsloth-bnb-4bit",
    "unsloth/LFM2-350M-unsloth-bnb-4bit",

    # Full 16bit unquantized models
    "unsloth/LFM2-1.2B",
    "unsloth/LFM2-700M",
    "unsloth/LFM2-350M",
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastModel.from_pretrained(
    model_name = "unsloth/LFM2-1.2B",
    dtype = None, # None for auto detection
    max_seq_length = 2048, # Choose any for long context!
    load_in_4bit = False,  # 4 bit quantization to reduce memory
    full_finetuning = False, # [NEW!] We have full finetuning now!
    # token = "hf_...", # use one if using gated models
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.


    PyTorch 2.6.0+cu124 with CUDA 1204 (you have 2.9.0+cu126)
    Python  3.12.9 (you have 3.12.12)
  Please reinstall xformers (see https://github.com/facebookresearch/xformers#installing-xformers)
  Memory-efficient attention, SwiGLU, sparse and more won't be available.
  Set XFORMERS_MORE_DETAILS=1 for more details


Switching to PyTorch attention since your Xformers is broken.

Unsloth: Xformers was not installed correctly.
Please install xformers separately first.
Then confirm if it's correctly installed by running:
python -m xformers.info

Longer error message:
xFormers can't load C++/CUDA extensions. xFormers was built for:
    PyTorch 2.6.0+cu124 with CUDA 1204 (you have 2.9.0+cu126)
    Python  3.12.9 (you have 3.12.12)
  Please reinstall xformers (see https://github.com/facebookresearch/xformers#installing-xformers)
  Memory-efficient attention, SwiGLU, sparse and more won't be available.
🦥 Unsloth Zoo will now patch everything to make training faster!
==((====))==  Unsloth 2025.11.3: Fast Lfm2 patching. Transformers: 4.55.4.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.9.0+cu126. CUDA: 7.5. CUDA Toolkit: 12.6. Triton: 3.5.0
\        /    Bfloat16 = FALSE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/u

model.safetensors:   0%|          | 0.00/2.34G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/160 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/434 [00:00<?, ?B/s]

chat_template.jinja:   0%|          | 0.00/209 [00:00<?, ?B/s]

We now add LoRA adapters so we only need to update a small amount of parameters!

In [9]:
model = FastModel.get_peft_model(
    model,
    finetune_vision_layers     = False, # LFM for now is just text only
    finetune_language_layers   = True,  # Should leave on!
    finetune_attention_modules = True,  # Attention good for GRPO
    finetune_mlp_modules       = True,  # SHould leave on always!

    r = 16,           # Larger = higher accuracy, but might overfit
    lora_alpha = 16,  # Recommended alpha == r at least
    lora_dropout = 0,
    bias = "none",
    random_state = 3407,
)

Unsloth: Making `model.base_model.model.model` require gradients


<a name="Data"></a>
### Data Prep
We now use the `LFM` format for conversation style finetunes. We use [miriad/miriad-4.4M](https://huggingface.co/datasets/miriad/miriad-4.4M) dataset and we will add conversations style into the dataset. LFM renders multi turn conversations like below:

```
<|startoftext|><|im_start|>user
Hello!<|im_end|>
<|im_start|>assistant
Hey there!<|im_end|>
```

In [10]:
tokenizer.apply_chat_template([
    {"role" : "user", "content" : "Hello!"},
    {"role" : "assistant", "content" : "Hey there!"}
], tokenize = False)

'<|startoftext|><|im_start|>user\nHello!<|im_end|>\n<|im_start|>assistant\nHey there!<|im_end|>\n'

We get the first 10000 rows of the dataset

In [11]:
from datasets import load_dataset, Dataset

# 1. Stream the dataset
dataset = load_dataset("PleIAs/SYNTH", split="train", streaming=True)

# 2. Filter only German
def is_german(example):
    lang = example.get("language", "")
    return lang and "de" in lang.lower()

de_stream = dataset.filter(is_german)

# 3. Transform to conversations format
def transform(example):
    q = example.get("query", "").strip()
    a = example.get("synthetic_answer", "").strip()
    return {
        "conversations": [
            {"role": "user", "content": q},
            {"role": "assistant", "content": a}
        ]
    }

# 4. Define generator *function* (not generator object)
def gen():
    count = 0
    for ex in de_stream:
        yield transform(ex)
        count += 1
        if count >= 10000:   # limit to 10000 if desired
            break

# 5. Build the in-memory dataset
de_synth = Dataset.from_generator(gen)

print(de_synth)
print(de_synth[0])

Resolving data files:   0%|          | 0/500 [00:00<?, ?it/s]

Resolving data files:   0%|          | 0/500 [00:00<?, ?it/s]

Dataset({
    features: ['conversations'],
    num_rows: 10000
})
{'conversations': [{'content': 'Wie genau interagieren die Wärmegradienten zwischen dem kalten nordwestlichen Wintermonsun und dem warmen Sommerwind mit der Topographie der japanischen Inseln, um diese speziellen Kármán-Wirbelwolken im Meer zu erzeugen?', 'role': 'user'}, {'content': 'Die Wärmegradienten zwischen dem kalten nordwestlichen Wintermonsun und dem warmen Sommerwind spielen eine entscheidende Rolle bei der Entstehung der charakteristischen Kármán-Wirbelwolken über dem Japanischen Meer.\n\nIm Winter bringt der starke nordwestliche Monsun mit Geschwindigkeiten von 12-15 m/s kalte, trockene kontinentale Luft aus Richtung Sibirien. Diese Luftmasse führt zu markanten Temperaturunterschieden zwischen den kalten nördlichen Gebieten mit durchschnittlichen Lufttemperaturen von -20°C im Januar/Februar und den wärmeren südlichen Regionen mit etwa 5°C. Der Wintermonsun verstärkt zudem die Verdunstung und Oberflächenwasser

We now use `standardize_data_formats` to try converting datasets to the correct format for finetuning purposes!

In [12]:
from unsloth.chat_templates import standardize_data_formats
dataset = standardize_data_formats(de_synth)

Unsloth: Standardizing formats (num_proc=2):   0%|          | 0/10000 [00:00<?, ? examples/s]

Let's see how row 100 looks like!

In [13]:
dataset[100]

{'conversations': [{'content': 'Hey, stimmt\'s dass der Ursprung von "forest" im Lateinischen tatsächlich unklar ist, oder hab ich da was falsch im Kopf?',
   'role': 'user'},
  {'content': 'Ja, das stimmt tatsächlich - der Ursprung von "forest" im Lateinischen ist sehr unsicher und es gibt verschiedene konkurrierende Erklärungen für seine Herkunft.\n\nDas Wort "forest" leitet sich vom Altfranzösischen "forest" (auch "forès") ab, was "Wald, weites von Bäumen bedecktes Gebiet" bedeutete. Dieses wurde wahrscheinlich über das germanische Sprachgebiet - vermutlich über Fränkisch oder Althochdeutsch - aus dem mittelalterlichen Lateinischen "foresta" entlehnt, das "offenes Holz" bezeichnete.\n\nDer genaue Ursprung dieses mittelalterlichen Lateinischen "foresta" bleibt jedoch unklar. Es gibt zwei Haupttheorien:\n\nTheorie A: Das Wort stammt von der späten lateinischen Wendung "forestam silvam" ab, was "das äußere Holz" bedeutet.\n\nTheorie B: "foresta" war eine Latinisierung des fränkischen W

We now have to apply the chat template for `LFM` onto the conversations, and save it to `text`. We also remove the BOS token otherwise we'll get double BOS tokens!

In [14]:
def formatting_prompts_func(examples):
    texts = tokenizer.apply_chat_template(
        examples["conversations"],
        tokenize = False,
        add_generation_prompt = False,
    )
    return { "text" : [x.removeprefix(tokenizer.bos_token) for x in texts] }

dataset = dataset.map(formatting_prompts_func, batched = True)

Map:   0%|          | 0/10000 [00:00<?, ? examples/s]

In [15]:
dataset[0]["text"]

'<|im_start|>user\nWie genau interagieren die Wärmegradienten zwischen dem kalten nordwestlichen Wintermonsun und dem warmen Sommerwind mit der Topographie der japanischen Inseln, um diese speziellen Kármán-Wirbelwolken im Meer zu erzeugen?<|im_end|>\n<|im_start|>assistant\nDie Wärmegradienten zwischen dem kalten nordwestlichen Wintermonsun und dem warmen Sommerwind spielen eine entscheidende Rolle bei der Entstehung der charakteristischen Kármán-Wirbelwolken über dem Japanischen Meer.\n\nIm Winter bringt der starke nordwestliche Monsun mit Geschwindigkeiten von 12-15 m/s kalte, trockene kontinentale Luft aus Richtung Sibirien. Diese Luftmasse führt zu markanten Temperaturunterschieden zwischen den kalten nördlichen Gebieten mit durchschnittlichen Lufttemperaturen von -20°C im Januar/Februar und den wärmeren südlichen Regionen mit etwa 5°C. Der Wintermonsun verstärkt zudem die Verdunstung und Oberflächenwasserkonvektion bis in Tiefen von 30 Metern.\n\nIm Sommer hingegen kehrt der Wind 

**Test the model before Training**

Now let's test the model before training.

In [28]:
messages = [{
    "role": "user",
    "content": """Wie beantragt man Bürgergeld?"""
}]
inputs = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True, # Must add for generation
    return_tensors = "pt",
    tokenize = True,
    return_dict = True,
).to("cuda")

from transformers import TextStreamer
_ = model.generate(
    **inputs,
    max_new_tokens = 2048, # Increase for longer outputs!
    # Recommended Liquid settings!
    temperature = 0.3, min_p = 0.15, repetition_penalty = 1.05,
    streamer = TextStreamer(tokenizer, skip_prompt = True),
)

Um Bürgergeld zu beantragen, müssen Sie in der Regel folgende Schritte unternehmen:

1. **Kontakt mit der zuständigen Behörde**: Zunächst sollten Sie sich an die örtliche Sozialbehörde wenden, die für die Bearbeitung von Bürgergeldanträgen zuständig ist. Dies kann je nach Bundesland und Kommune variieren.

2. **Antragsformular ausfüllen**: Sie benötigen ein Antragsformular, das Sie entweder im Büro der Sozialbehörde ausfüllen oder online herunterladen können.

3. **Dokumente vorbereiten**: Sammeln Sie alle relevanten Unterlagen, die Ihre Situation belegen könnten. Dazu gehören zum Beispiel:
   - Personalausweis oder Reisepass
   - Nachweis über das Einkommen (z.B. Lohnsteuerbescheinigung)
   - Nachweis über die Ausgaben (z.B. Mietvertrag, Rechnungen)
   - Nachweis über die Abhängigkeit vom Empfänger (z.B. Pflegeverordnung, Erziehungsverantwortung)

4. **Antrag einreichen**: Geben Sie den ausgefüllten Antrag und die erforderlichen Dokumente bei der Sozialbehörde ein. Dies kann persönlic

<a name="Train"></a>
### Train the model
Now let's train our model. We do 60 steps to speed things up, but you can set `num_train_epochs=1` for a full run, and turn off `max_steps=None`.

In [29]:
from trl import SFTTrainer, SFTConfig
trainer = SFTTrainer(
    model = model,
    tokenizer = tokenizer,
    train_dataset = dataset,
    eval_dataset = None, # Can set up evaluation!
    args = SFTConfig(
        dataset_text_field = "text",
        per_device_train_batch_size = 2,
        gradient_accumulation_steps = 4, # Use GA to mimic batch size!
        warmup_steps = 5,
        num_train_epochs = 1, # Set this for 1 full training run.
        #max_steps = 100,
        learning_rate = 2e-5, # Reduce to 2e-5 for long training runs
        logging_steps = 1,
        optim = "adamw_8bit",
        weight_decay = 0.01,
        lr_scheduler_type = "linear",
        seed = 3407,
        report_to = "none", # Use this for WandB etc
    ),
)

Unsloth: Tokenizing ["text"] (num_proc=6):   0%|          | 0/10000 [00:00<?, ? examples/s]

We also use Unsloth's `train_on_completions` method to only train on the assistant outputs and ignore the loss on the user's inputs. This helps increase accuracy of finetunes!

In [30]:
from unsloth.chat_templates import train_on_responses_only
trainer = train_on_responses_only(
    trainer,
    instruction_part = "<|im_start|>user\n",
    response_part = "<|im_start|>assistant\n",
)

Map (num_proc=6):   0%|          | 0/10000 [00:00<?, ? examples/s]

Let's verify masking the instruction part is done! Let's print the 100th row again.

In [31]:
tokenizer.decode(trainer.train_dataset[100]["input_ids"])

'<|startoftext|><|im_start|>user\nHey, stimmt\'s dass der Ursprung von "forest" im Lateinischen tatsächlich unklar ist, oder hab ich da was falsch im Kopf?<|im_end|>\n<|im_start|>assistant\nJa, das stimmt tatsächlich - der Ursprung von "forest" im Lateinischen ist sehr unsicher und es gibt verschiedene konkurrierende Erklärungen für seine Herkunft.\n\nDas Wort "forest" leitet sich vom Altfranzösischen "forest" (auch "forès") ab, was "Wald, weites von Bäumen bedecktes Gebiet" bedeutete. Dieses wurde wahrscheinlich über das germanische Sprachgebiet - vermutlich über Fränkisch oder Althochdeutsch - aus dem mittelalterlichen Lateinischen "foresta" entlehnt, das "offenes Holz" bezeichnete.\n\nDer genaue Ursprung dieses mittelalterlichen Lateinischen "foresta" bleibt jedoch unklar. Es gibt zwei Haupttheorien:\n\nTheorie A: Das Wort stammt von der späten lateinischen Wendung "forestam silvam" ab, was "das äußere Holz" bedeutet.\n\nTheorie B: "foresta" war eine Latinisierung des fränkischen Wo

Now let's print the masked out example - you should see only the answer is present:

In [32]:
tokenizer.decode([tokenizer.pad_token_id if x == -100 else x for x in trainer.train_dataset[100]["labels"]]).replace(tokenizer.pad_token, " ")

'                                         Ja, das stimmt tatsächlich - der Ursprung von "forest" im Lateinischen ist sehr unsicher und es gibt verschiedene konkurrierende Erklärungen für seine Herkunft.\n\nDas Wort "forest" leitet sich vom Altfranzösischen "forest" (auch "forès") ab, was "Wald, weites von Bäumen bedecktes Gebiet" bedeutete. Dieses wurde wahrscheinlich über das germanische Sprachgebiet - vermutlich über Fränkisch oder Althochdeutsch - aus dem mittelalterlichen Lateinischen "foresta" entlehnt, das "offenes Holz" bezeichnete.\n\nDer genaue Ursprung dieses mittelalterlichen Lateinischen "foresta" bleibt jedoch unklar. Es gibt zwei Haupttheorien:\n\nTheorie A: Das Wort stammt von der späten lateinischen Wendung "forestam silvam" ab, was "das äußere Holz" bedeutet.\n\nTheorie B: "foresta" war eine Latinisierung des fränkischen Wortes "*forhist" (Wald, bewaldetes Land), das von fränkischen Schreibern an "forestam silvam" angepasst wurde - entsprechend der üblichen Praxis ihre

In [33]:
# @title Show current memory stats
gpu_stats = torch.cuda.get_device_properties(0)
start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)
print(f"GPU = {gpu_stats.name}. Max memory = {max_memory} GB.")
print(f"{start_gpu_memory} GB of memory reserved.")

GPU = Tesla T4. Max memory = 14.741 GB.
3.436 GB of memory reserved.


In [34]:
trainer_stats = trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 10,000 | Num Epochs = 1 | Total steps = 1,250
O^O/ \_/ \    Batch size per device = 2 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (2 x 4 x 1) = 8
 "-____-"     Trainable parameters = 9,142,272 of 1,179,482,880 (0.78% trained)


In [35]:
# @title Show final memory and time stats
used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
used_memory_for_lora = round(used_memory - start_gpu_memory, 3)
used_percentage = round(used_memory / max_memory * 100, 3)
lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)
print(f"{trainer_stats.metrics['train_runtime']} seconds used for training.")
print(
    f"{round(trainer_stats.metrics['train_runtime']/60, 2)} minutes used for training."
)
print(f"Peak reserved memory = {used_memory} GB.")
print(f"Peak reserved memory for training = {used_memory_for_lora} GB.")
print(f"Peak reserved memory % of max memory = {used_percentage} %.")
print(f"Peak reserved memory for training % of max memory = {lora_percentage} %.")

3804.8513 seconds used for training.
63.41 minutes used for training.
Peak reserved memory = 6.656 GB.
Peak reserved memory for training = 3.22 GB.
Peak reserved memory % of max memory = 45.153 %.
Peak reserved memory for training % of max memory = 21.844 %.


<a name="Inference"></a>
### Inference
Let's run the model! You can change the instruction and input - leave the output blank!



In [36]:
messages = [{
    "role": "user",
    "content": """Wie beantragt man Bürgergeld?""",
}]
inputs = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True, # Must add for generation
    return_tensors = "pt",
    tokenize = True,
    return_dict = True,
).to("cuda")

from transformers import TextStreamer
_ = model.generate(
    **inputs,
    max_new_tokens = 2048, # Increase for longer outputs!
    # Recommended Liquid settings!
    temperature = 0.3, min_p = 0.15, repetition_penalty = 1.05,
    streamer = TextStreamer(tokenizer, skip_prompt = True),
)

Die Antragstellung auf Bürgergeld erfolgt über das Sozialamt. Das Sozialamt prüft die Anträge auf die Berechtigung zur Auszahlung von Bürgergeld und trifft dabei auf verschiedene Kriterien.

Für die Antragstellung müssen Sie sich an das Sozialamt wenden, das Ihre Anträge bearbeitet. Die genauen Anforderungen und der Antragsprozess können je nach Region variieren.

Das Sozialamt prüft dabei verschiedene Faktoren wie Einkommen, Vermögen, Beschäftigungsstatus und andere finanzielle Situation. Dabei werden auch die gesetzlichen Regelungen und die individuellen Umstände Ihres Falles berücksichtigt.

Bürgergeld ist eine Form der sozialen Unterstützung für Menschen mit geringem Einkommen oder besonderen Bedürfnissen. Die Höhe des Betrags hängt von verschiedenen Faktoren ab, einschließlich des Einkommens, der Familiengröße und der Lebenshaltungskosten in Ihrer Region.

Für detaillierte Informationen zu den Antragsverfahren und den spezifischen Anforderungen in Ihrem Land empfehle ich Ihnen, di

In [37]:
messages = [{
    "role": "user",
    "content": """What are some factors that determine the prognosis of patients undergoing bowel resection for colorectal cancer?""",
}]
inputs = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True, # Must add for generation
    return_tensors = "pt",
    tokenize = True,
    return_dict = True,
).to("cuda")

from transformers import TextStreamer
_ = model.generate(
    **inputs,
    max_new_tokens = 128, # Increase for longer outputs!
    # Recommended Liquid settings!
    temperature = 0.3, min_p = 0.15, repetition_penalty = 1.05,
    streamer = TextStreamer(tokenizer, skip_prompt = True),
)

The prognosis of patients undergoing bowel resection for colorectal cancer is influenced by several key factors, which can be broadly categorized into patient-related, tumor-related, and surgical-related factors.

Patient-related factors include age, performance status, and comorbidities. Younger patients with good overall health tend to have better outcomes. Performance status, often assessed using the Karnofsky or Eastern Cooperative Oncology Group (ECOG) scales, correlates strongly with survival rates. Patients with higher performance status generally have improved prognoses.

Tumor-related factors encompass the stage and histological subtype of the colorectal cancer. Early-stage tumors (I and II) with localized disease


<a name="Save"></a>
### Saving, loading finetuned models
To save the final model as LoRA adapters, either use Huggingface's `push_to_hub` for an online save or `save_pretrained` for a local save.

**[NOTE]** This ONLY saves the LoRA adapters, and not the full model. To save to 16bit or GGUF, scroll down!

In [38]:
from google.colab import userdata
token = userdata.get('secretName')

model.save_pretrained("lora_model")  # Local saving
tokenizer.save_pretrained("lora_model")
model.push_to_hub("furukama/lfm2_synth1000", token = token) # Online saving
tokenizer.push_to_hub("furukama/lfm2_synth1000", token = token) # Online saving

HfHubHTTPError: 401 Client Error: Unauthorized for url: https://huggingface.co/api/repos/create (Request ID: Root=1-6924e09a-0cb972301a774ee90fc9ee77;1a25e2c9-0ec9-471b-ae1f-dd60ea5a4f8c)

Invalid username or password.

Now if you want to load the LoRA adapters we just saved for inference, set `False` to `True`:

In [ ]:
if False:
    from unsloth import FastModel
    from transformers import Lfm2ForCausalLM
    model, tokenizer = FastModel.from_pretrained(
        model_name = "lora_model", # YOUR MODEL YOU USED FOR TRAINING
        max_seq_length = max_seq_length,
        dtype = dtype,
        auto_model = Lfm2ForCausalLM,
        load_in_4bit = load_in_4bit,
    )
    FastModel.for_inference(model) # Enable native 2x faster inference

messages = [{
    "role": "user",
    "content": "How do I code up a transformer?",
}]
inputs = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True, # Must add for generation
    return_tensors = "pt",
    tokenize = True,
    return_dict = True,
).to("cuda")

from transformers import TextStreamer
_ = model.generate(
    **inputs,
    max_new_tokens = 128, # Increase for longer outputs!
    # Recommended Liquid settings!
    temperature = 0.3, min_p = 0.15, repetition_penalty = 1.05,
    streamer = TextStreamer(tokenizer, skip_prompt = True),
)

To code up a transformer, you can follow these steps:

1. Import the necessary libraries:
```python
import torch
from transformers import TFAutoModelForSequenceClassification, TFAutoModelForTokenClassification
```

2. Load the pre-trained model and tokenizer:
```python
model_name = "bert-base-uncased"
model = TFAutoModelForSequenceClassification.from_pretrained(model_name)
tokenizer = TFAutoModelForTokenClassification.from_pretrained(model_name)
```

3. Define the input and output classes:
```python
input_class


You can also use Hugging Face's `AutoModelForPeftCausalLM`. Only use this if you do not have `unsloth` installed. It can be hopelessly slow, since `4bit` model downloading is not supported, and Unsloth's **inference is 2x faster**.

In [ ]:
if False:
    # I highly do NOT suggest - use Unsloth if possible
    from peft import AutoPeftModelForCausalLM
    from transformers import AutoTokenizer
    model = AutoPeftModelForCausalLM.from_pretrained(
        "lora_model", # YOUR MODEL YOU USED FOR TRAINING
        load_in_4bit = load_in_4bit,
    )
    tokenizer = AutoTokenizer.from_pretrained("lora_model")

### Saving to float16 for VLLM

We also support saving to `float16` directly. Select `merged_16bit` for float16 or `merged_4bit` for int4. We also allow `lora` adapters as a fallback. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [ ]:
# Merge to 16bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_16bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_16bit", token = "")

# Merge to 4bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_4bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_4bit", token = "")

# Just LoRA adapters
if False:
    model.save_pretrained("model")
    tokenizer.save_pretrained("model")
if False:
    model.push_to_hub("hf/model", token = "")
    tokenizer.push_to_hub("hf/model", token = "")

### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now! We clone `llama.cpp` and we default save it to `q8_0`. We allow all methods like `q4_k_m`. Use `save_pretrained_gguf` for local saving and `push_to_hub_gguf` for uploading to HF.

Some supported quant methods (full list on our [Wiki page](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)):
* `q8_0` - Fast conversion. High resource use, but generally acceptable.
* `q4_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q4_K.
* `q5_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q5_K.

[**NEW**] To finetune and auto export to Ollama, try our [Ollama notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)

In [ ]:
# Save to 8bit Q8_0
if False: model.save_pretrained_gguf("model", tokenizer,)
# Remember to go to https://huggingface.co/settings/tokens for a token!
# And change hf to your username!
if False: model.push_to_hub_gguf("hf/model", tokenizer, token = "")

# Save to 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "f16", token = "")

# Save to q4_k_m GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "q4_k_m")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "q4_k_m", token = "")

# Save to multiple GGUF options - much faster if you want multiple!
if False:
    model.push_to_hub_gguf(
        "hf/model", # Change hf to your username!
        tokenizer,
        quantization_method = ["q4_k_m", "q8_0", "q5_k_m",],
        token = "",
    )

Now, use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp.

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
